hera_qm/metrics_io.py

```diff
@@ -0,0 +1,640 @@
+                key_set = grp.create_dataset('key_order', data=key_order)
+
+                # Add boolean attribute to determine if key is a string
+                # Used to parse keys saved to dicts when reading
+                key_set.attrs['key_is_string'] = True
+            grp.attrs['key_is_string'] = isinstance(key, str)
+            _recursively_save_dict_to_group(h5file, path + key_str + '/',
+                                            in_dict[key])
+        else:
+            raise TypeError("Cannot save key {0} with type {1}"
+                            .format(key, type(in_dict[key])))
+    return
+
+
+def _recursively_make_dict_arrays_strings(in_dict):
+    """Recursively search dict for numpy array and cast as string.
+
+    Numpy arrays by default are space delimited in strings, this makes them comma delimitedself.
+
+    Arguments
+        in_dict: Dictionary to recursively search for numpy array
+    Returns
+        out_dict: Copy of in_dict with numpy arrays cast as comma delimited strings
+    """
+    out_dict = {}
+    for key in in_dict:
+        key_str = str(key)
+        if isinstance(in_dict[key], dict):
+            out_dict[key_str] = _recursively_make_dict_arrays_strings(in_dict[key])
+        elif isinstance(in_dict[key], np.ndarray):
+            out_dict[key_str] = np.array2string(in_dict[key],  separator=',')
+        else:
+            out_dict[key_str] = str(in_dict[key])
+    return out_dict
+
+
+def write_metric_file(filename, input_dict):
+    """Convert the input dictionary into an HDF5 File.
+
+    Can write either HDF5 (recommended) or JSON (Depreciated in Future) types.
+    Will try to guess which type based on the extension of the input filename.
+    If not extension is given, HDF5 is used by default.
+
+    If writing to HDF5 the following attributes will be added
+        Datasets:
+            key_is_string: Boolean flag to determine if dictionary key is a string
+
+        Groups:
+            group_is_ordered: Boolean flag to determine if dict was an OrderedDict
+
+
+    Arguments
+        filename: String of filename to which metrics will be written.
+                  Can include either HDF5 (recommended) or JSON (Depreciated in Future) extension.
+
+    Returns:
+        None
+    """
+    input_dict = copy.deepcopy(input_dict)
+    if filename.split('.')[-1] == 'json':
+        warnings.warn("JSON-type files can still be written "
+                      "but are no longer written by default.\n"
+                      "Write to HDF5 format for future compatibility.",
+                      PendingDeprecationWarning)
+
+        json_write_dict = _recursively_make_dict_arrays_strings(input_dict)
+
+        with open(filename, 'w') as outfile:
+            json.dump(json_write_dict, outfile, indent=4)
+    else:
+        if filename.split('.')[-1] not in ('hdf5', 'h5'):
+            filename += '.hdf5'
+
+        with h5py.File(filename, 'w') as f:
```

Comment:
This is interesting I hadn't thought about repopulating an existing file, as currently implemented it this will just blow away existing files with the new ones.

Reply:
So we could change the call on line 210 from 'w' to 'a' which will read/write to an existing file or create a file if none exists and try/except a bunch of requires. or 'w-' which would error if the file already exists and make the user supply like an overwrite or clobber keyword to replace it.